rules-testing/test.js

```diff
@@ -0,0 +1,313 @@
+            admin.settings({
+                host: "localhost:8000",
+                ssl: false
+            });
+            users = [{username: 'olly', uid: 'odog1'},
+                    {username: 'olance', uid: 'odog2'},
+                    {username: 'ol', uid: 'odog3'},
+                    {username: 'olsdafdsfdsafdsfsdafdsafdsafdsfasfdsa', uid: 'odog4'},
+                    {username: 'AAAAAA', uid: 'odog5'},
+                    {username: 'ofrog', uid: 'odog6'}];
+            for (let i = 0; i < users.length; i++){
+                await admin.collection("users")
+                .doc(users[i].uid)
+                .set({username: users[i].username, uid: users[i].uid})
+            }
+            usernames = [{username: 'olly', uid: 'odog1'}];
+            for(let i = 0; i < usernames.length; i++){
+                await admin.collection("usernames")
+                .doc(usernames[i].username)
+                .set({username: usernames[i].username, uid: usernames[i].uid})
+            }
+        })
+
+        it("Can't make username that is already used", async () => {
+            myAuth = {uid: 'odog1'};
+            setUpDatabase(myAuth);
+            const ref = db.collection("usernames");
+            await firebase.assertFails(ref.doc('olly')
+                .set({username: 'olly', uid: myAuth.uid}));
+        })
+
+        it("Can't make username where uid does not exist in users", async () => {
+            myAuth = {uid: 'odog7'};
+            setUpDatabase(myAuth);
+            const ref = db.collection("usernames");
+            await firebase.assertFails(ref.doc('olance')
+                .set({username: 'olance', uid: myAuth.uid}));
+        })
+
+        it("Can make username that isn't already used", async () => {
+            myAuth = {uid: 'odog2'};
+            setUpDatabase(myAuth);
+            const ref = db.collection("usernames");
+            await firebase.assertSucceeds(ref.doc('olance')
+                .set({username: 'olance', uid: myAuth.uid}));
+        })
+
+        //adds this to an actual user (even though that wouldn't be the case)
+        //to make sure it wouldn't add a username with too few characters
+        it("Won't make username with too few characters", async () => {
+            myAuth = {uid: 'odog3'};
+            setUpDatabase(myAuth);
+            const ref = db.collection("usernames");
+            await firebase.assertFails(ref.doc('ol')
+                .set({username: 'ol', uid: myAuth.uid}));
+        })
+
+        it("Won't make username with too many characters", async () => {
+            myAuth = {uid: 'odog4'};
+            setUpDatabase(myAuth);
+            const ref = db.collection("usernames");
+            await firebase.assertFails(ref.doc('olsdafdsfdsafdsfsdafdsafdsafdsfasfdsa')
+                .set({username: 'olsdafdsfdsafdsfsdafdsafdsafdsfasfdsa', uid: myAuth.uid}));
+        })
+
+        it("Won't make username with capitol letters", async () => {
```

Comment:
capital is for letters (among other things), capitol is for buildings  :P

Reply:
Spelling is not my strong suit haha!